contracts/wrapper/NameWrapper.sol

```diff
@@ -622,9 +632,14 @@ contract NameWrapper is
             CANNOT_TRANSFER | CANNOT_SET_RESOLVER | CANNOT_SET_TTL
         )
     {
-        ens.setRecord(node, address(this), resolver, ttl);
-        (address oldOwner, , ) = getData(uint256(node));
-        _transfer(oldOwner, owner, uint256(node), 1, "");
+        if (owner == address(0)) {
```

Comment:
It's the same behavior as `setSubnodeOwner()`, which unwraps when owner is set to 0. To be consistent, I added this behavior to setRecord so you could unwrap and burn a name but setting it to 0. I don't think it creates any bugs if we remove it, but it just means that you can have a wrapped name where the owner is 0. For `setSubnodeOwner()` this is definitely necessary, because settings the owner to 0 is how define deleting a subname. In this situation it would be effectively the same, but using the owner to setResolver/TTL and then unwrap, instead of calling `setRecord()` and then calling `unwrap()`.

The options I see are:

1) Leave as-is, owner can be burned to 0 with the NameWrapper still as the owner.
2) Implement the above behavior. Setting to 0, burns token (unwraps) and then sets the owner to 0 in the registry as if the domain has been deleted. Actually thinking about this now, I haven't accounted for .eth names, which would not be unwrapped.
3) Revert when setting the owner to 0

Reply:
I think `setSubnodeOwner` with an owner of 0 is the only way for a name owner to delete subdomains, right? So we can't eliminate that behaviour. I think you're right that this is the correct behaviour here too for consistency - though as you observe it also has to handle .eth 2LDs.

If we're going to start adding special code for .eth 2LDs that decodes the first label, is it viable for us to do away with all the special methods that work exclusively on .eth 2LDs and do that everywhere?

Reply:
That's interesting. I will have a think and explore that today

Reply:
Actually my first thought is: it only really works for these functions *because* we have the `IS_DOT_ETH` fuse. And that fuse is set when a name is wrapped. If we did away with `wrapETH2LD()` and had it use the `wrap` method instead. We'd have to do the string manipulation in solidity that increases gas by quite a bit (which you argued against), which is why we have the `IS_DOT_ETH` fuse

Reply:
What about `unwrapETH2LD` and `upgradeETH2LD`?

Reply:
I checked these. They both take parentNode and so there was nothing stopping us merging the logic together. It would just be an if statement and splitting the logic. I'm not sure it's worthwhile changing this logic given we made an explicit decision to split them in the past. For example:

`upgradeETH2LD` would call `wrapETH2LD `and upgrade would call `setSubnodeRecord`.

For `unwrapETH2LD` i don't think any of the current changes have changed any logic or made it simpler,.

Reply:
As discussed I've changed `setRecord()` to unwrap when set to 0 for subdomains or DNS names, and revert for .eth since we can't transfer to 0 for .eth